docs/ADR-3-Dependencies-version-constraints-in-cabal-file.md

```diff
@@ -0,0 +1,87 @@
+    * **Advantages**
+      * Fine grained control over what dependency is included where, with the single place which defines
+        version bounds.
+    * **Disadvantages**
+      * A lot of boilerplate when defining new dependencies.
+
+1. **Moving common dependencies to `common` stanzas, multiple dependencies in one stanza**.
+  This allows for grouping of dependencies between components in order specify version bound only once.
+  For example the same `common` stanza could be reused in `B` and `C` in order to define version bounds in a
+  single place.
+    * **Advantages**
+      * Grouping multiple dependencies together introduces less boilerplate in comparison to the
+        *one-stanza-per-one-dependency* option.
+    * **Disadvantages**
+      * It may be hard to define logical grouping for few unrelated dependencies.
+      * In the worst case scenario we'll end up with *one-stanza-per-one-dependency* definition of
+        dependencies, when those will be used in different components.
+
+1. **Declaring dependencies' bounds everywhere**.
+  In the provided example, the version bounds of common dependencies of `A` and `B` should be both defined in
+  `A` and `B`.
+    * **Advantages**
+      * Makes it easy to remember how to update dependency version everywhere.
+    * **Disadvantages**
+      * It's not hard to overlook one spot when a person updating dependency version is unfamiliar with the
+        rule.
+
+1. **Declaring dependencies' bounds once, as close as possible to the dependency tree root.**
+  In the provided example, the version bounds of common dependencies of `A` and `B` should be only defined in
+  `A`.
+  If dependencies are both present in `B` & `C`, the version bounds should be duplicated in both components.
+    * **Advantages**
+      * Slight de-duplication of version bounds definitions in comparison to duplicated version bounds
+      * Does not require introducing `common` stanzas
+    * **Disadvantages**
+      * In the `B` & `C` case, there will be a duplication of version bounds.
+      * Inconsistency in version bounds declaration because of `B` & `C` case.
+
+1. **Using a configuration language allowing to generate cabal files** like https://github.com/dhall-lang/dhall-to-cabal .
+  This way instead of modifying cabal files, one would has to modify configuration file and then regenerate
+  `.cabal` files from those.
+    * **Advantages**
+      * Very flexible way of configuration, allows for deduplication of cabal files' contents.
+    * **Disadvantages**
+      * New development tool with non-zero learning effort required for team members.
+      * Additional piece requiring integration in CI.
+      * Syntax usually is more verbose than just `.cabal` files itself.
+
+# Decision
+
+TBD
+
+# Consequences
+
+The adopted version constraints setting policy should be used consistently across all the projects.
```

Comment:
We need to be careful about imposing ADRs on all teams and projects unless they are included in the discussion.